app/src/main/java/co/condorlabs/customcomponents/helper/masks/PhoneNumberTextWatcherMask.kt

```diff
@@ -17,54 +17,95 @@
+ * @company Condor Labs.
+ * @email eduque@condorlabs.io.
+ */
+class PhoneNumberTextWatcherMask(
+    private val mask: String,
+    val cursorPos: (Int) -> Unit
+) : TextWatcher {
+    private var isRunning = false
+    private var isDeleting = false
+    private var specialCharacters = EMPTY
+    private val charToDescribeANumber = CHAR_TO_DESCRIBE_A_NUMBER
+    private var maskPos = ZERO
+    private var numPos = ZERO
+    private var start = ZERO
+    private var after = ZERO
 
-    override fun afterTextChanged(s: Editable?) {
-        s?.let { text ->
-            receiver.removeTextChangedListener(this)
-            var result = text.toString().replace(PHONE_NUMBER_SEPARATOR_TOKEN, "")
+    init {
+        specialCharacters = mask.replace(charToDescribeANumber.toString(), EMPTY)
+    }
 
-            when (result.length) {
-                in PHONE_NUMBER_REGEX_FIRST_GROUP_RANGE_BOTTOM..PHONE_NUMBER_REGEX_FIRST_GROUP_RANGE_TOP -> {
-                    result = result.replaceFirst(
-                        PHONE_NUMBER_REGEX_FIRST_AND_SECOND_GROUP_MATCHER.toRegex(),
-                        PHONE_NUMBER_REGEX_FIRST_GROUP_REPLACEMENT_MATCHER
-                    )
-                }
-                in PHONE_NUMBER_REGEX_SECOND_GROUP_RANGE_BOTTOM..PHONE_NUMBER_REGEX_SECOND_GROUP_RANGE_TOP -> {
-                    result = result.replaceFirst(
-                        "$PHONE_NUMBER_REGEX_FIRST_AND_SECOND_GROUP_MATCHER$PHONE_NUMBER_REGEX_FIRST_AND_SECOND_GROUP_MATCHER".toRegex(),
-                        PHONE_NUMBER_REGEX_SECOND_GROUP_REPLACEMENT_MATCHER
-                    )
-                }
-                in PHONE_NUMBER_REGEX_THIRD_GROUP_RANGE_BOTTOM..PHONE_NUMBER_REGEX_THIRD_GROUP_RANGE_TOP -> {
-                    result = result.replaceFirst(
-                        ("$PHONE_NUMBER_REGEX_FIRST_AND_SECOND_GROUP_MATCHER" +
-                                "$PHONE_NUMBER_REGEX_FIRST_AND_SECOND_GROUP_MATCHER" +
-                                "$PHONE_NUMBER_REGEX_THIRD_GROUP_MATCHER").toRegex(),
-                        PHONE_NUMBER_REGEX_THIRD_GROUP_REPLACEMENT_MATCHER
-                    )
+    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
+        if (!isRunning) {
+            maskPos = ZERO
+            numPos = ZERO
+            this.start = start
+            this.after = after
+            isDeleting = count > after
+        }
+    }
+
+    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
+
+    override fun afterTextChanged(editable: Editable?) {
+        if (isRunning) return
+
+        val text = editable?.toString() ?: return
+        val numbers = textWithoutSpecialCharacters(text, specialCharacters)
+        val finalText = StringBuilder()
+        var finalSelection = after
+
+        if (numbers.isEmpty()) {
+            isRunning = true
+            editable.clear()
+            isRunning = false
```

Comment:
Why `isRunning` is changing in line 68 too?

Reply:
any changes made on `editable` will call the watcher again, `isRunning` is used to allow or not the changes on it.